lib/puppet/parser/functions.rb

```diff
@@ -98,10 +98,20 @@ def self.get_function_info(name)
         def self.add_function_info(name, info)
           @metadata[name] = info
         end
+
+        # This is triggered if user calls def self.some_method inside the body given to newfunction
+        def singleton_method_added(m)
+          raise SecurityError, _("Illegal method definition of method '%{method_name}' in legacy function") % { method_name: m }
```

Comment:
Is there anything that can be added to this error message (or documentation the user can be pointed to) about what they should be doing instead of trying to define the method here?

Reply:
A docs page on the handful of better ways to do it would be incredibly helpful. As I understand it, there are basically three ways:

* refactor to use a proc/lamda or something that doesn't require defining named function
* refactor to use a shared library class and define the function there
* refactor to a Puppet 4.x function and leave 3.x support in the past. (recommended)

Quick examples of these methods in the docs would be rad.

Reply:
Point me to where you want to see the docs and I can write that up.

Reply:
@hlindberg , you could give me docs either in the predocs repo or in a Google doc. To add the docs, I'd also need a ticket with some context, and if you have ideas, where this information should live in the docs. Thank you!

Reply:
Added predocs in a google doc: https://docs.google.com/document/d/1_KToXDDjbe0HQ6I4UqckNqcJ-AHreH_dmd_A1Cr85jU/edit# will link that to the ticket as well

Reply:
I'd like to see this error message updated to include a pointer to wherever the docs are going to end up, but other than that, I'm 👍 on the PR.